### PR TITLE
feat: csrf protection for login endpoints

### PR DESCRIPTION
**Pull request by wgwz**
## Description

Closes: https://github.com/regen-network/regen-registry/issues/1486

See the task above for a pretty thorough description of why I think we need to add this.
Also see the task above for some reasoning on the third-party library I chose to implement CSRF protection with.

Particularly useful to get started with this review, is this code, which shows exactly how to use an HTTP client to successfully pass CSRF protections:

https://github.com/regen-network/registry-server/blob/cdb16ca0bfd5a09e85e9adc303917084b1038783/server/__tests__/utils.ts#L12-L22

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] targeted `dev` branch
- [ ] provided a link to the relevant issue or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed
- [ ] submit a PR against `master` branch after this one (and other PRs depending on this one, e.g. on regen-network/regen-web, if applicable) get merged

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please **add
your handle next to the items reviewed if you only reviewed selected items**.*

I have...

- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed code correctness and readability
- [ ] reviewed documentation is accurate
- [ ] reviewed tests
- [ ] manually tested (if applicable)

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -30,14 +31,15 @@
...
     "@types/jest": "^27.4.1",
     "@types/node": "^17.0.21",
+    "@types/node-fetch": "^2.6.2",
     "@typescript-eslint/eslint-plugin": "^5.13.0",
     "@typescript-eslint/parser": "^5.13.0",
     "eslint": "^8.10.0",
     "eslint-config-prettier": "^8.5.0",
     "eslint-plugin-prettier": "^4.0.0",
-    "jest": "^26.6.3",
+    "jest": "^29",
```
Comment (wgwz) [inline package.json:40]: this bump was needed to fix an error with jest when trying to import third-party libraries into testing code, it's an error related to ESM or something. the newer versions of jest better handle the issue. it can be reproduced in `master` by adding an `axios` import into a jest test file.

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+WORKDIR /app
+COPY yarn.lock .
+COPY package.json .
+COPY lerna.json .
+COPY common/package.json common/package.json
+COPY iri-gen/package.json iri-gen/package.json
+COPY server/package.json server/package.json
+RUN yarn
+COPY . .
+CMD (cd server && yarn build && yarn start)
```
Comment (wgwz) [inline Dockerfile:12]: this is a change that i thought would make the CI test setup in GH easier but it's not necessary yet. see the `registryServer.yml` file for how we're currently running the newly added e2e tests. we need to have a running server in CI for the e2e tests.

File: .github/workflows/registryServer.yml [hunk 3]
```diff
@@ -63,4 +63,16 @@ jobs:
...
+        env:
+          MAILERLITE_API_KEY: foobar
+          AIRTABLE_API_KEY: foobar
+          AUTH0_DOMAIN: foobar
+          AUTH0_CLIENT_ID: foobar
+          AUTH0_CLIENT_SECRET: foobar
+          DATABASE_URL: postgres://postgres:postgres@localhost:5432/regen_registry
+          CSRF_SECRET: foobar
+          CSRF_COOKIE_NAME: __Host-regen-dev.x-csrf-token
+        run: yarn serve &
```
Comment (wgwz) [inline .github/workflows/registryServer.yml:77]: we start up a registry server instance in the background of the CI job for running tests against. it works, but this could be flaky. i prefer to wait and see if this becomes a problem.

File: server/__tests__/e2e/web3auth.login.test.ts [hunk 4]
```diff
@@ -0,0 +1,17 @@
+import fetch from 'node-fetch';
+import { CSRFRequest } from '../utils';
+
+describe('web3auth login endpoint', () => {
```
Comment (wgwz) [inline server/__tests__/e2e/web3auth.login.test.ts:4]: beginning a test module for the web3auth login endpoints. more tests will be added here in a separate PR.

File: server/__tests__/utils.ts [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+export async function fetchCsrf(): Promise<{ cookie: string; token: string }> {
+  const resp = await fetch('http://localhost:5000/csrfToken', {
+    method: 'GET',
+  });
+  const cookie = resp.headers.get('set-cookie');
+  const { token } = await resp.json();
+  return { cookie, token };
+}
+
+export async function CSRFRequest(endpoint: string): Promise<Request> {
```
Comment (wgwz) [inline server/__tests__/utils.ts:12]: shows how to set up a POST request with double-csrf pattern

File: server/routes/web3auth.ts [hunk 6]
```diff
@@ -3,17 +3,23 @@ import passport from 'passport';
...
 
-web3auth.use('/login', passport.authenticate('keplr'), (req, res) => {
-  return res.send({
-    user: req.user,
-    message: 'You have been signed in via keplr!',
-  });
-});
+web3auth.use(
+  '/login',
+  doubleCsrfProtection,
```
Comment (wgwz) [inline server/routes/web3auth.ts:12]: protect the login and logout endpoints (see below)

**Review by blushi — COMMENTED** [review]
looks good, just a few nits

File: server/Server.ts [hunk 7]
```diff
@@ -291,6 +297,9 @@ app.use((err, req, res, next) => {
   console.error('req info:', { params, query, body, path });
   next(err);
 });
+
+import { invalidCsrfTokenError } from './middleware/csrf';
```
[the remark this answers is not in the document]
Comment (wgwz) [reply server/Server.ts]: i installed that eslint plugin and ran it for our codebase, it introduced a few fixes after i ran it with `--fix` via e194fd245f8e360c3771cf10bb8e8a9a47368a35

File: server/routes/csrf.ts [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+export const csrfRouter = express.Router();
+
+csrfRouter.get('/csrfToken', (req, res) => {
+  const token = generateToken(res, req);
+  return res.send({ token });
+});
+
+csrfRouter.post('/csrfToken', doubleCsrfProtection, (req, res) => {
+  console.log(req);
+  console.log(res);
```
[the remark this answers is not in the document]
Comment (wgwz) [reply server/routes/csrf.ts]: removed!
